### ISSUE TITLE
InProcess hosting model connection issue

### ISSUE THREAD

**Issue by sepehr1014** [aspnet/AspNetCore#17065]
### Describe the bug
Using the InProcess hosting model in production, simultaneous connections are capped to around 100 (we usually have more than a thousand connected web sockets) and we repeatedly get exceptions about connection issues similar to the one below.

```
Microsoft.AspNetCore.Connections.ConnectionResetException: The client has disconnected
 ---> System.Runtime.InteropServices.COMException (0x800704CD): An operation was attempted on a nonexistent network connection. (0x800704CD)
   --- End of inner exception stack trace ---
   at Microsoft.AspNetCore.Server.IIS.Core.IO.AsyncIOOperation.GetResult(Int16 token)
   at Microsoft.AspNetCore.Server.IIS.Core.IISHttpContext.ReadBody()
   at System.IO.Pipelines.PipeCompletion.ThrowLatchedException()
   at System.IO.Pipelines.Pipe.GetReadResult(ReadResult& result)
   at System.IO.Pipelines.Pipe.ReadAsync(CancellationToken token)
   at System.IO.Pipelines.Pipe.DefaultPipeReader.ReadAsync(CancellationToken cancellationToken)
   at Microsoft.AspNetCore.Server.IIS.Core.IISHttpContext.ReadAsync(Memory`1 memory, CancellationToken cancellationToken)
   at Microsoft.AspNetCore.Server.IIS.Core.HttpRequestStream.ReadAsyncInternal(Memory`1 buffer, CancellationToken cancellationToken)
   at Microsoft.AspNetCore.WebUtilities.BufferedReadStream.EnsureBufferedAsync(Int32 minCount, CancellationToken cancellationToken)
   at Microsoft.AspNetCore.WebUtilities.MultipartReaderStream.ReadAsync(Byte[] buffer, Int32 offset, Int32 count, CancellationToken cancellationToken)
   at Microsoft.AspNetCore.WebUtilities.StreamHelperExtensions.DrainAsync(Stream stream, ArrayPool`1 bytePool, Nullable`1 limit, CancellationToken cancellationToken)
   at Microsoft.AspNetCore.WebUtilities.MultipartReader.ReadNextSectionAsync(CancellationToken cancellationToken)
   at Pasaj.MultipartProviderHelper.ExtractModelAsync[T](HttpRequest request, ILogger`1 logger) in C:\Web\Patogh\Patogh\Helpers\MultipartProviderHelperNew.cs:line 130
   at Pasaj.MultipartModelBinder`1.BindModelAsync(ModelBindingContext bindingContext)
```

### To Reproduce
Use InProcess hosting model in production with thousands of users. This does not occur in OutOfProcess mode. The same issue occurs in 2.2 (or any other version), just with different exception types.

### Further technical details
- ASP.NET Core 3.0
- Using IIS on Windows Server 2016

**Comment by jkotalik** [comment]
@davidfowl

**Comment by Tratcher** [comment]
> The same issue occurs in 2.2 (or any other version), just with different exception types.

What are the 2.2 exceptions?

**Comment by sepehr1014** [comment]
> > The same issue occurs in 2.2 (or any other version), just with different exception types.
> 
> What are the 2.2 exceptions?

Unfortunately I don't have the logs anymore. We've migrated to 3.0 and had to switch to OutOfProcess mode. Though that has some others issues.

**Comment by jkotalik** [comment]
@sepehr1014 we will try our best to repro it, @Tratcher may contact you for more infromation.

**Comment by jkotalik** [comment]
Also, how did you get the number of 100 connections?

**Comment by sepehr1014** [comment]
@jkotalik Thanks.

We have a middleware that handles web socket requests and adds the `WebSocket `object along with session info to a `ConcurrentDictionary`. The session is removed upon disconnection. So at any given time, the `Count `of this `ConcurrentDictionary `represents the number of online users. When using InProcess that number stays around 100 while we usually have at least 1,000 online users.

BTW we also provide RESTful APIs as you can see from that exception log.

We're also setting these options for `ConfigureWebHostDefaults() `in `CreateHostBuilder `:

```C#
webBuilder.ConfigureKestrel(serverOptions =>
                    {
                        // Set properties and call methods on options
                        serverOptions.Limits.MinRequestBodyDataRate = new MinDataRate(12, TimeSpan.FromMinutes(2));
                        serverOptions.Limits.KeepAliveTimeout = TimeSpan.FromMinutes(20);
                        serverOptions.Limits.MaxConcurrentConnections = 8192;
                        serverOptions.Limits.MaxConcurrentUpgradedConnections = 32768;
                    })
```

I see the new default for MaxConcurrentConnections is unlimited so we can skip them. But the issue at hand shouldn't be related to these.

**Comment by davidfowl** [comment]
Those limits only apply to kestrel not IIS

**Comment by sepehr1014** [comment]
@davidfowl Are there any IIS settings I should tweak? This issue is only present when using InProcess hosting model.

**Comment by sepehr1014** [comment]
@Tratcher No, it seems to be fixed. I'm not sure if it's fixed by upgrading to 3.1.0 or it was IIS config issue:

https://serverfault.com/questions/992954/http-sys-limitexceeded-error/993450

**Comment by anurse** [comment]
> No, it seems to be fixed

I'll close this then, since it seems to be resolved.
